staging/src/k8s.io/apiserver/pkg/util/panic/panic.go

```diff
@@ -0,0 +1,198 @@
+	// value and corresponding stack trace
+	// The request handler uses this to wait for the spawned
+	// goroutine to finish
+	Done() <-chan interface{}
+}
+
+func NewInvoker() *invoker {
+	return &invoker{ch: make(chan interface{}, 1)}
+}
+
+// FormatPanicValueWithStackTrace formats the given panic value
+// specified in 'recovered' to an error object.
+// If the given panic value is nil, then the function return nil.
+// The returned error object includes the runtime stack
+// trace from the calling goroutine.
+// If the panic value is the sentinel ErrAbortHandler
+// then the sentinel is returned as is.
+func FormatPanicValueWithStackTrace(recovered interface{}) interface{} {
+	if recovered == nil {
+		return nil
+	}
+	// do not wrap the sentinel ErrAbortHandler panic value
+	if recovered == http.ErrAbortHandler {
+		return recovered
+	}
+
+	// Same as stdlib http server code. Manually allocate stack
+	// trace buffer size to prevent excessively large logs
+	const size = 64 << 10
+	buf := make([]byte, size)
+	buf = buf[:runtime.Stack(buf, false)]
+	return &recoveredPanicError{
+		reason:     recovered,
+		stackTrace: buf,
+	}
+}
+
+// WrapRecoveredPanic wraps the recovered panic value given in
+// 'own' and the given panic value in 'other'.
+// - own is the recovered panic value from the caller's goroutine
+// - other is the recovered panic value from the other goroutine
+// via a receiving channel in the caller's goroutine
+// If both the caller goroutine and the other/spawned
+// goroutine panic, then:
+// a) the spawned goroutine should use FormatPanicValueWithStackTrace
+// to convert the panic value to an error object and then return
+// it to the caller goroutine via the channel
+// b) The parent goroutine should use WrapRecoveredPanic to wrap its own
+// recovered panic value and the panic value from the spawned goroutine
+func WrapRecoveredPanic(own, other interface{}) interface{} {
+	switch {
+	// both the caller goroutine and the spawned goroutine panicked
+	case own != nil && other != nil:
+		return &wrappedPanicError{
+			own:   own,
+			other: other,
+		}
+	// only the spawned goroutine panicked, no need to wrap
+	case other != nil:
+		return other
+	// only the caller goroutine panicked, no need to wrap
+	case own != nil:
+		return own
+	default:
+		return nil
+	}
+}
+
+type invoker struct {
+	ch chan interface{}
+}
+
+func (i *invoker) Done() <-chan interface{} { return i.ch }
+
+func (i *invoker) InvokeWithPanicProtection(f func()) {
+	defer func() {
+		var err interface{}
+		if recovered := recover(); recovered != nil {
+			err = FormatPanicValueWithStackTrace(recovered)
+		}
+		// so the request handler can wait until the spawned
+		// goroutine is done
+		i.ch <- err
+	}()
+	f()
+}
+
+func (i *invoker) InvokeWithPanicWrapping(f func()) {
+	defer func() {
+		own := recover()
+		other := <-i.ch
+		if err := WrapRecoveredPanic(own, other); err != nil {
+			// both goroutines have panicked
```

Comment:
s/both/one or both/

Reply:
fixed in https://github.com/kubernetes/kubernetes/pull/127075